fastrand.go

```diff
@@ -42,36 +49,52 @@ func init() {
 
 // Read fills b with random data. It always returns len(b), nil.
 func (r *randReader) Read(b []byte) (int, error) {
-	// Update the progress and use the result + the seed entropy to form a
-	// unique seed to hash.
-	progress := atomic.AddUint64(&r.progress, 1)
-
-	// If it seems like the progress is about to reset, increment the first
-	// entropy bytes. Without this check, the cycle time is 2^64 iterations, but
-	// with the check it is a secure 2^128 iterations.
-	if progress > 1<<63 {
-		atomic.AddUint64((*uint64)(unsafe.Pointer(&r.entropy[0])), 1)
-		atomic.StoreUint64(&r.progress, 0)
+	// Grab a unique counter from the reader, while atomically updating the
+	// counter so that concurrent callers also end up with unique values.
+	counter := atomic.AddUint64(&r.counter, 1)
+	counterExtra := atomic.LoadUint64(&r.counterExtra)
+	// Update the second 64 bits of the counter if the first 64 bits are close
+	// to wrapping around. It is possible that the second 64 bits of the counter
+	// is updates multiple times by several concurrent threads. This wastes part
+	// of the counter space (up to 2^63 items each time), however the overall
+	// space is large enough (2^128) that wasting some every reset does not make
+	// it any more likely that the caller exhaust the whole possible search
+	// space.
+	if counter > 1<<63 {
+		atomic.AddUint64(&r.counterExtra, 1)
+		atomic.StoreUint64(&r.counter, 0)
```

Comment:
Using `==` doesn't work because one thread could get `counter = 1<<63+1, extra = 1` when extra flips from 0 to 1 (because the thread grabs counter, and then extra increments, and then it grabs extra) and then on the next round you have a thread grabbing `counter = 1 << 63+1, extra =1` when the counter is flipping from 1 to 2. So you get overlaps, except this time they are extremely far apart and much less likely to be detected or disruptive.

But it's still an issue.